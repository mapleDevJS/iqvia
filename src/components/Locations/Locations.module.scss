.locations {
    display: flex;
    flex-direction: column;
    min-height: 80vh;
    padding: 1rem;
    border-radius: 0.5rem;
    box-shadow: var(--shadow);

    @media (max-width: 70rem) {
        min-height: auto;
    }
}

.searchBox {
    position: relative;
    display: flex;
    justify-content: space-between;
}

.inputLocation {
    width: 100%;

    margin-right: 1rem;
    padding: 0.5rem 0.5rem 0.5rem 0;

    border: 0;
    border: 2px solid var(--transparent);
    border-bottom: 2px solid var(--border-light);

    &:focus-visible {
        border-bottom: 2px solid var(--border-dark);
        outline: none;
    }
}

.title {
    margin: 1.25rem 0 0 0;
    font-weight: 500;
    font-weight: 500;
    font-size: 1rem;
    color: var(--font-dark);
}

.bottomControls {
    display: flex;
    justify-content: flex-end;
    margin-top: auto;
    padding-top: 1rem;
}

.notFound {
    position: absolute;
    bottom: -1.5rem;
    font-weight: 500;
    font-size: 0.75rem;
    color: var(--error-primary);
}
