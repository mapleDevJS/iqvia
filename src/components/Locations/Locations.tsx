import { ChangeEvent, useContext, useState } from 'react';
import { observer } from 'mobx-react';
import { store } from 'store/index';
import RecentLocations from './RecentLocations/RecentLocations';
import Button from 'components/Button/Button';
import styles from './Locations.module.scss';

enum ButtonText {
    ADD = 'Add',
    CLEAR = 'Clear',
}

const Locations = () => {
    const { locationsStore } = useContext(store);
    const { fetchLocations, getLocations, resetLocations, resetForecast, getError, resetError } =
        locationsStore;

    const [cityName, setCityName] = useState<string>('');

    const inputChangeHandler = (evt: ChangeEvent<HTMLInputElement>) => {
        setCityName(evt.target.value);
    };

    const onKeyPressHandler = (evt: React.KeyboardEvent) => {
        if (evt.key === 'Enter') {
            setCityName('');
            fetchLocations(cityName);
        }
    };

    const error = getError();
    const locations = getLocations();
    const isAnyLocation = locations.length > 0;

    const addButtonClickHandler = (cityName: string) => {
        setCityName('');
        fetchLocations(cityName);
    };

    const clearButtonClickHandler = () => {
        resetForecast();
        resetLocations();
    };

    return (
        <section className={styles.locations}>
            <div className={styles.searchBox}>
                <input
                    className={styles.inputLocation}
                    placeholder={'Type city name'}
                    value={cityName}
                    onKeyPress={onKeyPressHandler}
                    onChange={inputChangeHandler}
                    onFocus={resetError}
                />
                <Button onClick={() => addButtonClickHandler(cityName)} text={ButtonText.ADD} />
                {error && <span className={styles.notFound}>{error.message}</span>}
            </div>

            <h2 className={styles.title}>Recent Locations</h2>

            {isAnyLocation && <RecentLocations locations={locations} />}

            <div className={styles.bottomControls}>
                <Button onClick={clearButtonClickHandler} text={ButtonText.CLEAR} />
            </div>
        </section>
    );
};

export default observer(Locations);
