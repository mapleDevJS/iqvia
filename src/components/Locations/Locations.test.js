import userEvent from '@testing-library/user-event';
import { render, screen, cleanup } from '@testing-library/react';
import Locations from './Locations';

afterEach(cleanup);

describe('when rendered', () => {
    test('renders Locations component', () => {
        render(<Locations />);
    });

    it('should contain an expected input', () => {
        render(<Locations />);
        const inputElement = screen.getByRole('textbox');
        expect(inputElement).toBeInTheDocument();

    });

    it('should contain an expected input with a placeholder', () => {
        render(<Locations />);
        const placeholderElement = screen.getByPlaceholderText('Type city name');
        expect(placeholderElement).toBeInTheDocument();

    });

    it('should contain an expected header', () => {
        render(<Locations />);
        const headerElement = screen.getByText('Recent Locations');
        expect(headerElement).toBeInTheDocument();
    });

    it('should contain an expected add button', () => {
        render(<Locations />);
        const buttonElement = screen.getByText('Add');
        expect(buttonElement).toBeInTheDocument();
    });

    it('should contain an expected clear button', () => {
        render(<Locations />);
        const buttonElement = screen.getByText('Clear');
        expect(buttonElement).toBeInTheDocument();
    });
});

describe('when typed in a City Name input', () => {
    it('should update its value', () => {
        render(<Locations />);

        const inputElement = screen.getByPlaceholderText(/Type city name/);

        userEvent.clear(inputElement);
        userEvent.type(inputElement, 'Toronto');
        expect(inputElement).toHaveValue('Toronto');
    });

    it('should clear its value', () => {
        render(<Locations />);

        const inputElement = screen.getByPlaceholderText(/Type city name/);

        userEvent.clear(inputElement);
        userEvent.type(inputElement, 'Toronto{enter}');
        expect(inputElement).toHaveValue('');
    });
});

describe('snapshot', () => {
    it('should take a snapshot', () => {
        const { asFragment } = render(<Locations />)
        expect(asFragment(<Locations />)).toMatchSnapshot()
    })
});
