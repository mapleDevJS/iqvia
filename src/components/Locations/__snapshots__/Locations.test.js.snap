// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot should take a snapshot 1`] = `
<DocumentFragment>
  <section
    class="locations"
  >
    <div
      class="searchBox"
    >
      <input
        class="inputLocation"
        placeholder="Type city name"
        value=""
      />
      <button
        class="button undefined"
      >
        Add
      </button>
    </div>
    <h2
      class="title"
    >
      Recent Locations
    </h2>
    <div
      class="bottomControls"
    >
      <button
        class="button undefined"
      >
        Clear
      </button>
    </div>
  </section>
</DocumentFragment>
`;
